### PR TITLE
Preserve adapter trainability during injection

### LINKED ISSUE #3487: Adapter injection applies a new adapter's inference mode to active adapters
### System Info

- PEFT `main` at `a5526d27a9d47d1e8264d5e1b1f96c0fdc79464e`
- PEFT 0.20.0
- Python 3.13.13
- PyTorch 2.13.0+cpu
- Transformers 5.14.1
- Windows 11

### Who can help?

@BenjaminBossan

### Reproduction

Four existing strict-xfail tests still reproduce the same adapter-specific trainability problem on current `main`:

```bash
pytest \
  tests/test_custom_models.py::TestRequiresGrad::test_loading_model_requires_grad_set_correctly_switch_inference_mode \
  tests/test_custom_models.py::TestRequiresGrad::test_loading_model_requires_grad_load_adapter_then_add_adapter \
  tests/test_custom_models.py::TestRequiresGrad::test_requires_grad_adalora_different_targets \
  tests/test_custom_models.py::TestRequiresGrad::test_requires_grad_adalora_same_targets \
  --runxfail -q
```

Result: 4 failed.

The failures are:

- Loading a frozen `default` adapter and then loading trainable `other` leaves the `default` parameters trainable instead of `other`.
- Adding another adapter to a model loaded with `is_trainable=False` re-enables gradients on the frozen adapter.
- Adding an AdaLoRA adapter with `inference_mode=True` freezes the still-active trainable adapter, for both same-target and different-target adapters.

These cases appear to share the housekeeping call in `BaseTuner.inject_adapter`:

```python
self.set_adapter(self.active_adapters, inference_mode=peft_config.inference_mode)
```

Here `peft_config` belongs to the adapter being injected, while `active_adapters` can refer to a different adapter. The new adapter's `inference_mode` is therefore applied across adapter boundaries.

If this scope is welcome, I can prepare one focused fix covering the load/add and AdaLoRA cases and turn the four existing xfails into regular regression tests. I would not add or change public API.

### Expected behavior

Injecting or loading an adapter should not change the trainability state of a different adapter. The active adapter should keep its current `requires_grad` state, while the newly loaded adapter should follow its own `is_trainable` or `inference_mode` setting.

### PR DESCRIPTION
**Pull request by DaoyuanLi2816**
## What does this PR do?

Fixes #3487 and #3580.

`BaseTuner.inject_adapter` reruns adapter activation during housekeeping. Without preserving the previous state, adding or loading an adapter can unexpectedly change `requires_grad` on existing base-model or adapter parameters.

This snapshots every pre-existing parameter's `requires_grad` value and restores it after adapter housekeeping and auxiliary-module setup. The first injection still follows normal PEFT initialization and freezes the base model. `PeftModel.load_adapter(..., is_trainable=True)` then explicitly enables only the newly loaded adapter; `add_adapter` leaves the new inactive adapter frozen. UniLoRA's trainable theta parameter is classified as an adapter layer so the generic helper handles it consistently.

Regression coverage now checks same and opposite trainability across PEFT configurations, the `add_adapter` path, modules-to-save and trainable-token cases, and preservation of manually customized base/adapter gradient state.

## Coordination

- Initial issue and approval: https://github.com/huggingface/peft/issues/3487#issuecomment-5142080203
- Real-world report and maintainer direction: https://github.com/huggingface/peft/issues/3580#issuecomment-5426486884

## Validation

- `pytest tests/test_custom_models.py -k requires_grad -q --no-cov --regression`: 209 passed, 4 skipped
- Ruff 0.15.12 check and format check on the changed files
- `git diff --check`

## AI assistance

AI assistance was used to inspect the failure, prepare the patch, and run validation. I reviewed every line of the final diff and the test results, understand the change end to end, and take responsibility for the contribution.

### CODE REVIEW

File: tests/test_custom_models.py [hunk 1]
```diff
@@ -6731,16 +6718,10 @@ def test_loading_model_requires_grad_set_correctly_switch_inference_mode(self, c
...
         assert params_with_grad == expected
 
-    @pytest.mark.xfail(strict=True)
-    @pytest.mark.parametrize("config_cls", [LoraConfig])  # no need to check each method, they all fail
+    @pytest.mark.parametrize("config_cls", [LoraConfig])  # no need to check each method
     def test_loading_model_requires_grad_load_adapter_then_add_adapter(self, config_cls, tmp_path):
-        # When adding a new adapter with model.add_adapter, through the set_adapter call in update_layer, we activate
-        # the gradients of the first adapter, even if it's not desired. Since there is no is_trainable argument on
-        # add_adapter, there is no way to disable that at the moment.
-        # When/If this is fixed, the check can be integrated into test_loading_model_requires_grad_set_correctly and
```
Comment (BenjaminBossan) [inline tests/test_custom_models.py:7039]: The comment mentions that this test should be folded into `test_loading_model_requires_grad_set_correctly` when the issue is fixed. Currently, that test only checks what happens if loading two adapters with the same value for `is_trainable`. Let's extend that test to load a third adapter with the opposite value, and then check for correctness.

File: src/peft/tuners/tuners_utils.py [hunk 2]
```diff
@@ -783,6 +783,18 @@ def inject_adapter(
         ###################################
         # PREPARATION OF MODEL AND CONFIG #
         ###################################
+        # Adding an adapter must not change the trainability of parameters that were already present. Some tuners
+        # create their new adapter parameters before calling this method, so exclude those from the snapshot.
+        existing_adapter_parameters = (
```
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py]: The name is a bit misleading, as this list contains both the parameter and the `requires_grad` status.

Moreover, I'm not too happy that we add two more loops here, which are not free: One over `model.modules()` and one over `model.named_parameters()`. Further below, we already have a loop over the modules: `named_modules = list(model.named_modules())`. That loop should be extended instead of adding new loops here.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for working on this. I have added some comments, please check.

File: tests/test_custom_models.py (hunk 1 above)
Comment (DaoyuanLi2816) [reply tests/test_custom_models.py:7039]: Updated the existing parametrized test to load a third adapter with the opposite is_trainable value and verify that only the intended adapter remains trainable. I also folded the state capture into the existing named_modules traversal and removed the redundant standalone tests. The focused test groups pass (68 cases each, with 2 expected skips in the modules-to-save group), and Ruff check/format pass.

Comment (DaoyuanLi2816) [reply tests/test_custom_models.py:7039]: Updated in a580cf97 to follow the later direction in #3580: the opposite-`is_trainable` case remains out of scope and the strict xfail is restored. The existing same-state regression coverage remains folded into the parametrized test.

**Comment by BenjaminBossan** [comment]
@DaoyuanLi2816 As always, please ping me once you're done with the updates and the PR is ready for review.

**Comment by BenjaminBossan** [comment]
I forgot to comment on this part:

> The remaining xfail loads a new adapter with `is_trainable=True` while the active adapter is frozen. Supporting that case would require `PeftModel.load_adapter` to track a newly added adapter and explicitly enable its parameters after loading, even though the adapter is not active. A narrower version could do so only when the model had no trainable parameters before the load. This changes the current relationship between activation and trainability, so it is intentionally left out of this PR.

This is not quite clear to me, why and how do we need to track the newly added adapter? Could we not use `BaseTuner.set_requires_grad`?

File: tests/test_custom_models.py (hunk 1 above)
Comment (DaoyuanLi2816) [reply tests/test_custom_models.py:7039]: Updated in 2c53a878 following the maintainer direction in #3580. The parametrized all-config test now covers the opposite-`is_trainable` load, and the implementation uses `BaseTuner.set_requires_grad` while preserving existing adapter state. The expanded focused suite passes (134 passed, 2 skipped), along with Ruff check/format and `git diff --check`.

@BenjaminBossan, this is ready for re-review.

File: tests/test_custom_models.py [hunk 3]
```diff
@@ -6616,6 +6603,17 @@ def test_loading_model_requires_grad_set_correctly(self, config_cls, is_trainabl
...
+        model.load_adapter(tmp_path, adapter_name="opposite", is_trainable=not is_trainable)
+        for name, param in model.named_parameters():
+            if skip_ranknum and "ranknum" in name:
+                continue
+            expected_trainable_adapter = "default" if is_trainable else "opposite"
+            if f".{expected_trainable_adapter}" in name:
+                assert param.requires_grad
+            else:
+                assert not param.requires_grad
+
```
Comment (BenjaminBossan) [inline tests/test_custom_models.py:6916]: This test does not quite cover the removed `test_loading_model_requires_grad_load_adapter_then_add_adapter` test yet, as that one calls `add_adapter`, which could theoretically work differently from `load_adapter` when it comes to setting `requires_grad`.

So I would suggest to add one more check here, namely calling `model.add_adapter` to add a 4th adapter. As `add_adapter` does not an `is_trainable` argument, it is implicitly expected to be `False`. So we should check that the 4th adapter has `requires_grad` and the previous adapters stay the same.

I know that this whole test is becoming quite large because of this. As a whole, the test could be simplified and cleaned up, some comments are also imprecise (for instance confounding "active" with "requires_grad"). But let's leave the clean up for the future.

File: tests/test_custom_models.py [hunk 4]
```diff
@@ -6616,6 +6603,17 @@ def test_loading_model_requires_grad_set_correctly(self, config_cls, is_trainabl
         else:
             assert all(not p.requires_grad for p in model.parameters())
 
+        # load a third adapter with the opposite trainability; it is not automatically activated
+        model.load_adapter(tmp_path, adapter_name="opposite", is_trainable=not is_trainable)
+        for name, param in model.named_parameters():
+            if skip_ranknum and "ranknum" in name:
+                continue
+            expected_trainable_adapter = "default" if is_trainable else "opposite"
```
Comment (BenjaminBossan) [inline tests/test_custom_models.py]: This assignment can be moved out of the loop.

File: src/peft/tuners/tuners_utils.py [hunk 5]
```diff
@@ -829,6 +829,25 @@ def inject_adapter(
...
+                existing_adapter_prefixes.append(key + ".")
+            for parameter_name, parameter in module.named_parameters(recurse=False):
+                full_name = f"{key}.{parameter_name}" if key else parameter_name
+                if id(parameter) in seen_parameters:
+                    continue
+                seen_parameters.add(id(parameter))
+                if f".{adapter_name}." not in full_name and not full_name.endswith(f".{adapter_name}"):
+                    existing_parameter_trainability.append((parameter, parameter.requires_grad))
+
+        if not existing_adapter_prefixes:
```
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py:894]: I'm trying to understand this: `existing_parameter_trainability` records the `requires_grad` of parameters that are _not_ directly related to the current PEFT adapter. Then, if there are no `existing_adapter_prefixes`, i.e. when this is the first adapter being added, we null `existing_parameter_trainability`. Why is that?

File: src/peft/tuners/tuners_utils.py [hunk 6]
```diff
@@ -829,6 +829,25 @@ def inject_adapter(
         named_modules = list(model.named_modules())
         key_list = [key for key, _ in named_modules]
 
+        # Adding an adapter must not change the trainability of parameters that were already present. Some tuners
+        # create their new adapter parameters before calling this method, so exclude those from the snapshot.
+        existing_adapter_prefixes = []
+        existing_parameter_trainability = []
```
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py]:
```suggestion
        mapping_existing_parameter_requires_grad = []
```

It's unwieldy, but let's be precise with naming, as the situation is already confusing enough. It's not a dict, but at least functionally, this is like a mapping.

File: src/peft/tuners/tuners_utils.py [hunk 7]
```diff
@@ -1056,10 +1070,13 @@ def inject_adapter(
...
 
+        for parameter, requires_grad in existing_parameter_trainability:
+            parameter.requires_grad = requires_grad
+
         if self.peft_config[adapter_name].inference_mode:
-            for n, p in model.named_parameters():
-                if adapter_name in n:
-                    p.requires_grad = False
+            self.set_requires_grad(adapter_name, requires_grad=False)
+        elif not any(requires_grad for _, requires_grad in existing_parameter_trainability):
```
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py]: This is not clear to me, could you please add a comment on what we cover here?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for working on this complex topic.

I think we should take this opportunity to try to think this through from first principles. There are many assumptions baked into how we deal with `requires_grad` right now and my reading is that this PR does not fully resolve them (yet).

Let's say someone creates a PEFT model with the "default" adapter. Then, _normally_, the base model parameters all have `requires_grad=False` and the adapter parameters have `requires_grad=True`. But let's say the user manually activates some gradients from the base model parameters and deactivates some from the adapter. After calling `load_adapter` or `add_adapter`, do we want to conserve those settings or do we want to reset them to the "correct" setting?

I lean toward not changing those values, even if they might not make sense. A naive solution would thus be to record _all_ `requires_grad` values at the beginning of `inject_adapter` and after injection, we reset all `requires_grad` to these recorded values. For the newly added adapter, if it comes with `is_trainable=True`, we call `set_requires_grad(new_adapter_name, True)`, otherwise we don't (so it should have `requires_grad=False`). Most likely, I'm missing something here, but would that approach work?

For BC, we should still tie `requires_grad` to `set_adapter`, as discussed in an earlier thread, but that shouldn't affect this PR.

File: tests/test_custom_models.py (hunk 3 above)
Comment (DaoyuanLi2816) [reply tests/test_custom_models.py]: System.Collections.Hashtable.body

File: src/peft/tuners/tuners_utils.py (hunk 5 above)
Comment (DaoyuanLi2816) [reply src/peft/tuners/tuners_utils.py]: System.Collections.Hashtable.body

File: src/peft/tuners/tuners_utils.py (hunk 6 above)
Comment (DaoyuanLi2816) [reply src/peft/tuners/tuners_utils.py]: System.Collections.Hashtable.body

File: src/peft/tuners/tuners_utils.py (hunk 7 above)
Comment (DaoyuanLi2816) [reply src/peft/tuners/tuners_utils.py]: System.Collections.Hashtable.body

**Comment by DaoyuanLi2816** [comment]
@BenjaminBossan Updated in baf7db16 following the first-principles approach from your review: injection now restores every pre-existing parameter's custom `requires_grad` state after housekeeping, while `load_adapter(..., is_trainable=True)` explicitly enables the newly loaded adapter. I also added the `add_adapter` path and a regression for manually customized base/adapter gradients, and updated the same-trainability expectations.

`pytest tests/test_custom_models.py -k requires_grad -q --no-cov --regression` passes with 209 passed and 4 skipped; Ruff check/format and `git diff --check` pass. Ready for re-review.

File: tests/test_custom_models.py (hunk 3 above)
Comment (DaoyuanLi2816) [reply tests/test_custom_models.py:6916]: Added the fourth `add_adapter` check. It verifies that the new inactive adapter stays frozen while all previously established `requires_grad` states remain unchanged; AdaLoRA is skipped because it structurally rejects adding another training-mode config.

File: tests/test_custom_models.py (hunk 4 above)
Comment (DaoyuanLi2816) [reply tests/test_custom_models.py]: Moved `expected_trainable_adapters` out of the parameter loop.

File: src/peft/tuners/tuners_utils.py (hunk 5 above)
Comment (DaoyuanLi2816) [reply src/peft/tuners/tuners_utils.py:894]: Added a comment clarifying this case: on the first injection, the base model should be frozen, so we deliberately discard the snapshot instead of restoring the base parameters to their default trainability.

File: src/peft/tuners/tuners_utils.py (hunk 6 above)
Comment (DaoyuanLi2816) [reply src/peft/tuners/tuners_utils.py]: Renamed it to `mapping_existing_parameter_requires_grad` as suggested.

File: src/peft/tuners/tuners_utils.py (hunk 7 above)
Comment (DaoyuanLi2816) [reply src/peft/tuners/tuners_utils.py]: Removed this special-case branch. `inject_adapter` now only restores pre-existing parameter state; `load_adapter(..., is_trainable=True)` explicitly enables the newly loaded adapter with `set_requires_grad`, and the custom-state regression covers the distinction.

File: tests/test_custom_models.py [hunk 8]
```diff
@@ -6606,16 +6593,58 @@ def test_loading_model_requires_grad_set_correctly(self, config_cls, is_trainabl
...
+        model.add_adapter(adapter_name="added", peft_config=config)
+        for name, param in model.named_parameters():
+            if skip_ranknum and "ranknum" in name:
+                continue
+            if any(f".{adapter_name}" in name for adapter_name in expected_trainable_adapters):
+                assert param.requires_grad
+            else:
+                assert not param.requires_grad
+
+    def test_loading_adapter_preserves_custom_requires_grad(self, tmp_path):
```
Comment (BenjaminBossan) [inline tests/test_custom_models.py:6931]: Let's add a short comment here on what use case this test covers.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the latest update, the PR shaping up quite well. Could you please merge the latest main branch so that `make style` can pass?

File: tests/test_custom_models.py (hunk 8 above)
Comment (DaoyuanLi2816) [reply tests/test_custom_models.py:6931]: Added a concise comment explaining that the test emulates a user manually enabling one base-model weight while freezing one adapter parameter.

**Comment by DaoyuanLi2816** [comment]
@BenjaminBossan Merged current `main` in 8bf74685 and added the requested use-case comment. The focused `requires_grad` suite passes with 214 passed and 4 skipped; Ruff check/format and diff check pass. Ready for re-review.

**Review by BenjaminBossan — APPROVED** [review]
Thanks for the last push. The PR LGTM now. I hope that this covers most of the remaining issues there still were with trainability. It is a change to the existing behavior, but it should be in line with what the user actually intended, so I'd consider this a bug fix and not a backwards incompatible change. I'll make sure to mention this in the release text for the next release.
